src/net.cpp

```diff
@@ -1543,30 +1552,55 @@ void CConnman::ThreadDNSAddressSeed()
         seeds_right_now = seeds.size();
     }
 
+    // goal: only query DNS seed if address need is acute
+    // * If we have a reasonable number of peers in addrman, spend
+    //   some time trying them first. This improves user privacy by
+    //   creating fewer identifying DNS requests, reduces trust by
+    //   giving seeds less influence on the network topology, and
+    //   reduces traffic to the seeds.
+    // * When querying DNS seeds query a few at once, this ensures
+    //   that we don't give DNS seeds the ability to eclipse nodes
+    //   that query them.
+    // * If we continue having problems, eventually query all the
+    //   DNS seeds, and if that fails too, also try the fixed seeds.
+    //   (done in ThreadOpenConnections)
     for (const std::string& seed : seeds) {
-        // goal: only query DNS seed if address need is acute
-        // Avoiding DNS seeds when we don't need them improves user privacy by
-        // creating fewer identifying DNS requests, reduces trust by giving seeds
-        // less influence on the network topology, and reduces traffic to the seeds.
         if (addrman.size() > 0 && seeds_right_now == 0) {
-            if (!interruptNet.sleep_for(std::chrono::seconds(11))) return;
-
-            LOCK(cs_vNodes);
-            int nRelevant = 0;
-            for (const CNode* pnode : vNodes) {
-                nRelevant += pnode->fSuccessfullyConnected && !pnode->fFeeler && !pnode->fOneShot && !pnode->m_manual_connection && !pnode->fInbound;
-            }
-            if (nRelevant >= 2) {
-                LogPrintf("P2P peers available. Skipped DNS seeding.\n");
-                return;
+            LogPrintf("Waiting %d seconds before querying DNS seeds.\n", seeds_wait_time.count());
+            std::chrono::seconds to_wait = seeds_wait_time;
+            while (to_wait.count() > 0) {
+                std::chrono::seconds w{25};
+                if (w > to_wait) w = to_wait;
+                if (!interruptNet.sleep_for(w)) return;
+                to_wait -= w;
+
+                int nRelevant = 0;
+                {
+                    LOCK(cs_vNodes);
+                    for (const CNode* pnode : vNodes) {
+                        nRelevant += pnode->fSuccessfullyConnected && !pnode->fFeeler && !pnode->fOneShot && !pnode->m_manual_connection && !pnode->fInbound;
+                    }
+                }
+                if (nRelevant >= 2) {
+                    if (found > 0) {
+                        LogPrintf("%d addresses found from DNS seeds\n", found);
+                        LogPrintf("P2P peers available. Finished DNS seeding.\n");
+                    } else {
+                        LogPrintf("P2P peers available. Skipped DNS seeding.\n");
+                    }
+                    return;
+                }
             }
             seeds_right_now += DNSSEEDS_TO_QUERY_AT_ONCE;
+        } else if (seeds_right_now == 0) {
```

Comment:
Not sure about this, I would say that's a behavior change. Previously node with  an empty addrman would have query all DNS seeds due to seeds_right_now == -1 as you noticed in #15558. Now you may exit early, if among DNSSEEDS_TO_QUERY_AT_ONCE you find at least 2 outbound peers but with final result of having less address origin diversity in your addrman..